### PR TITLE
[SPARK-17241][SparkR][MLlib] SparkR spark.glm should have configurable regularization parameter

### PR DESCRIPTION
**Pull request by keypointt**
https://issues.apache.org/jira/browse/SPARK-17241
## What changes were proposed in this pull request?

Spark has configurable L2 regularization parameter for generalized linear regression. It is very important to have them in SparkR so that users can run ridge regression.
## How was this patch tested?

Test manually on local laptop.

### CODE REVIEW

File: R/pkg/R/mllib.R [hunk 1]
```diff
@@ -171,7 +172,8 @@ predict_internal <- function(object, newData) {
 #' @note spark.glm since 2.0.0
 #' @seealso \link{glm}, \link{read.ml}
 setMethod("spark.glm", signature(data = "SparkDataFrame", formula = "formula"),
-          function(data, formula, family = gaussian, tol = 1e-6, maxIter = 25, weightCol = NULL) {
+          function(data, formula, family = gaussian, tol = 1e-6, regParam = 0.0, maxIter = 25,
+                   weightCol = NULL) {
```
Comment (junyangq) [inline R/pkg/R/mllib.R]: Perhaps we can add that to the end of the argument list so that it doesn't break the existing calls to the function?

Comment (keypointt) [inline R/pkg/R/mllib.R]: check the `fit()` method of the wrapper, as long as the parameter order matches, it's ok.

I've tested it already in R terminal.

Comment (junyangq) [inline R/pkg/R/mllib.R]: If say an R user call the function by `spark.glm(df, label ~ feature, gaussian, 1e-6, 25)`. This will break their code.

Comment (felixcheung) [inline R/pkg/R/mllib.R]: +1 - we should try to avoid breaking existing caller

File: R/pkg/inst/tests/testthat/test_mllib.R [hunk 2]
```diff
@@ -99,6 +99,10 @@ test_that("spark.glm summary", {
   expect_match(out[2], "Deviance Residuals:")
   expect_true(any(grepl("AIC: 59.22", out)))
 
+  # Test spark.glm works with regularization parameter
+  regStats <- summary(spark.glm(training, Sepal_Width ~ Sepal_Length + Species, regParam = 0.3))
+  expect_equal(regStats$aic, 136.7, tolerance = 1e-3)
```
Comment (junyangq) [inline R/pkg/inst/tests/testthat/test_mllib.R]: How was this number computed?

Comment (keypointt) [inline R/pkg/inst/tests/testthat/test_mllib.R]: oh I just check the output of model stats

maybe there is a better way to test it?

Comment (junyangq) [inline R/pkg/inst/tests/testthat/test_mllib.R]: I remember it should match the result of `glmnet`? Perhaps you can try the same example there or take a look at https://github.com/apache/spark/blob/master/mllib/src/test/scala/org/apache/spark/ml/regression/GeneralizedLinearRegressionSuite.scala#L307

Comment (junyangq) [inline R/pkg/inst/tests/testthat/test_mllib.R]: though it's very likely that the result would not change :)

**Comment by junyangq** [comment]
Other than that, LGTM.

**Comment by felixcheung** [comment]
LGTM - I guess it would be good to sanity check the test value with Scala.

**Comment by shivaram** [comment]
Thanks @keypointt for the PR and @junyangq @felixcheung for reviewing. Merging this into master

**Pull request closed without merging**